File: app/src/main/java/org/dhis2/usescases/login/LoginActivity.java
package org.dhis2.usescases.login;

import android.content.Intent;
import android.os.Bundle;
import android.os.PersistableBundle;
import android.text.InputType;
import android.view.View;
import android.view.WindowManager;
import android.widget.ArrayAdapter;

import com.andrognito.pinlockview.PinLockListener;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import org.dhis2.App;
import org.dhis2.R;
import org.dhis2.databinding.ActivityLoginBinding;
import org.dhis2.usescases.general.ActivityGlobalAbstract;
import org.dhis2.usescases.main.MainActivity;
import org.dhis2.usescases.sync.SyncActivity;
import org.dhis2.utils.BiometricStorage;
import org.dhis2.utils.ColorUtils;
import org.dhis2.utils.Constants;
import org.dhis2.utils.D2ErrorUtils;
import org.dhis2.utils.NetworkUtils;
import org.dhis2.utils.OnDialogClickListener;
import org.dhis2.utils.TestingCredential;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.io.StringWriter;
import java.io.Writer;
import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import androidx.appcompat.app.AlertDialog;
import androidx.core.content.ContextCompat;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.ViewModelProviders;
import de.adorsys.android.securestoragelibrary.SecurePreferences;
import timber.log.Timber;

import static android.text.TextUtils.isEmpty;
import static org.dhis2.utils.Constants.RQ_QR_SCANNER;


public class LoginActivity extends ActivityGlobalAbstract implements LoginContracts.View {

    ActivityLoginBinding binding;

    @Inject
    LoginContracts.Presenter presenter;

    List<String> users;
    List<String> urls;

    private boolean isPinScreenVisible = false;
    private String qrUrl;
    private LoginViewModel loginViewModel;
    private List<TestingCredential> testingCredentials = new ArrayList<>();


    @Override
    protected void onCreate(Bundle savedInstanceState) {

        LoginComponent loginComponent = ((App) getApplicationContext()).loginComponent();
        if (loginComponent == null) {
            // in case if we don't have cached presenter
            loginComponent = ((App) getApplicationContext()).createLoginComponent();
        }
        loginComponent.inject(this);

        super.onCreate(savedInstanceState);
        loginViewModel = ViewModelProviders.of(this).get(LoginViewModel.class);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_login);
        binding.setPresenter(presenter);
        binding.setLoginModel(loginViewModel);

        loginViewModel.isDataComplete().observe(this, this::setLoginVisibility);
        loginViewModel.isTestingEnvironment().observe(this, testingEnvironment -> {
            binding.serverUrlEdit.setText(testingEnvironment.val0());
            binding.userNameEdit.setText(testingEnvironment.val1());
            binding.userPassEdit.setText(testingEnvironment.val2());
        });
        setTestingCredentials();
        setAutocompleteAdapters();

    }

    private void setTestingCredentials() {
        int testingCredentialsIdentifier = getResources().getIdentifier("testing_credentials", "raw", getPackageName());
        if (testingCredentialsIdentifier != -1) {
            Writer writer = new StringWriter();
            char[] buffer = new char[1024];
            try (InputStream is = getResources().openRawResource(testingCredentialsIdentifier)) {
                Reader reader = new BufferedReader(new InputStreamReader(is, "UTF-8"));
                int n;
                while ((n = reader.read(buffer)) != -1) {
                    writer.write(buffer, 0, n);
                }
            } catch (Exception e) {
                Timber.e(e);
            }
            testingCredentials = new Gson().fromJson(writer.toString(), new TypeToken<List<TestingCredential>>(){}.getType());
            loginViewModel.setTestingCredentials(testingCredentials);
        }
    }


    @Override
    protected void onResume() {
        super.onResume();
        presenter.init(this);
        NetworkUtils.isGooglePlayServicesAvailable(this);
    }

    @Override
    protected void onPause() {
        presenter.onDestroy();
        super.onPause();
    }

    @Override
    protected void onDestroy() {
        ((App) getApplicationContext()).releaseLoginComponent();
        super.onDestroy();
    }

    @Override
    public void showBiometricButton() {
        binding.biometricButton.setVisibility(View.VISIBLE);
    }

    @Override
    public void checkSecuredCredentials() {
        if (SecurePreferences.contains(Constants.SECURE_SERVER_URL) &&
                SecurePreferences.contains(Constants.SECURE_USER_NAME) &&
                SecurePreferences.contains(Constants.SECURE_PASS)) {
            binding.serverUrlEdit.setText(SecurePreferences.getStringValue(Constants.SECURE_SERVER_URL, null));
            binding.userNameEdit.setText(SecurePreferences.getStringValue(Constants.SECURE_USER_NAME, null));
            binding.userPassEdit.setText(SecurePreferences.getStringValue(Constants.SECURE_PASS, null));
            showLoginProgress(true);
        } else
            showInfoDialog(getString(R.string.biometrics_dialog_title), getString(R.string.biometrics_first_use_text));
    }

    @Override
    public void goToNextScreen() {
        if (NetworkUtils.isOnline(this)) {
            startActivity(SyncActivity.class, null, true, true, null);
        } else
            startActivity(MainActivity.class, null, true, true, null);
    }

    @Override
    public void switchPasswordVisibility() {
        if (binding.userPassEdit.getInputType() == (InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_PASSWORD)) {
            binding.visibilityButton.setImageDrawable(
                    ColorUtils.tintDrawableWithColor(
                            ContextCompat.getDrawable(this, R.drawable.ic_visibility),
                            ColorUtils.getPrimaryColor(this, ColorUtils.ColorType.PRIMARY)));
            binding.userPassEdit.setInputType(InputType.TYPE_CLASS_TEXT);
        } else {
            binding.visibilityButton.setImageDrawable(
                    ColorUtils.tintDrawableWithColor(
                            ContextCompat.getDrawable(this, R.drawable.ic_visibility_off),
                            ColorUtils.getPrimaryColor(this, ColorUtils.ColorType.PRIMARY)));
            binding.userPassEdit.setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_PASSWORD);
        }

        binding.userPassEdit.setSelection(binding.userPassEdit.getText().length());
    }

    @Override
    public void setUrl(String url) {
        binding.serverUrlEdit.setText(!isEmpty(qrUrl) ? qrUrl : url);
    }

    @Override
    public void showUnlockButton() {
        binding.unlockLayout.setVisibility(View.VISIBLE);
    }

    @Override
    public void renderError(Throwable throwable) {
        showInfoDialog(getString(R.string.login_error), D2ErrorUtils.getErrorMessage(this, throwable));
    }

    @Override
    public void handleLogout() {
        recreate();
    }

    @Override
    public void setLoginVisibility(boolean isVisible) {
        binding.login.setVisibility(isVisible ? View.VISIBLE : View.GONE);
    }

    @Override
    public void showLoginProgress(boolean showLogin) {
        if (showLogin) {
            getWindow().setFlags(WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE,
                    WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE);
            binding.credentialLayout.setVisibility(View.GONE);
            binding.progressLayout.setVisibility(View.VISIBLE);

            presenter.logIn(
                    binding.serverUrl.getEditText().getText().toString(),
                    binding.userName.getEditText().getText().toString(),
                    binding.userPass.getEditText().getText().toString()
            );
        } else {
            getWindow().clearFlags(WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE);
            binding.credentialLayout.setVisibility(View.VISIBLE);
            binding.progressLayout.setVisibility(View.GONE);
        }
    }

    @Override
    public void showCrashlyticsDialog() {
        showInfoDialog(getString(R.string.send_user_name_title), getString(R.string.send_user_name_mesage),
                new OnDialogClickListener() {
                    @Override
                    public void onPossitiveClick(AlertDialog alertDialog) {
                        getSharedPreferences().edit().putBoolean(Constants.USER_ASKED_CRASHLYTICS, true).apply();
                        getSharedPreferences().edit().putString(Constants.USER, binding.userName.getEditText().getText().toString()).apply();
                        showLoginProgress(true);
                    }

                    @Override
                    public void onNegativeClick(AlertDialog alertDialog) {
                        getSharedPreferences().edit().putBoolean(Constants.USER_ASKED_CRASHLYTICS, true).apply();
                        showLoginProgress(true);
                    }
                }).show();
    }

    @Override
    public void onUnlockClick(View android) {
        binding.pinLayout.pinLockView.attachIndicatorDots(binding.pinLayout.indicatorDots);
        binding.pinLayout.pinLockView.setPinLockListener(new PinLockListener() {
            @Override
            public void onComplete(String pin) {
                presenter.unlockSession(pin);
            }

            @Override
            public void onEmpty() {

            }

            @Override
            public void onPinChange(int pinLength, String intermediatePin) {

            }
        });
        binding.pinLayout.getRoot().setVisibility(View.VISIBLE);
        isPinScreenVisible = true;
    }

    @Override
    public void onLogoutClick(View android) {
        presenter.logOut();
    }

    @Override
    public void setAutocompleteAdapters() {

        urls = getListFromPreference(Constants.PREFS_URLS);
        users = getListFromPreference(Constants.PREFS_USERS);

        if (!urls.contains(Constants.URL_TEST_229))
            urls.add(Constants.URL_TEST_229);
        if (!urls.contains(Constants.URL_TEST_230))
            urls.add(Constants.URL_TEST_230);
        if (!users.contains(Constants.USER_TEST_ANDROID))
            users.add(Constants.USER_TEST_ANDROID);

        for(TestingCredential testingCredential : testingCredentials){
            if(!urls.contains(testingCredential.getServer_url()))
                urls.add(testingCredential.getServer_url());
        }

        saveListToPreference(Constants.PREFS_URLS, urls);
        saveListToPreference(Constants.PREFS_USERS, users);

        ArrayAdapter<String> urlAdapter = new ArrayAdapter<>(this, android.R.layout.simple_dropdown_item_1line, urls);
        ArrayAdapter<String> userAdapter = new ArrayAdapter<>(this, android.R.layout.simple_dropdown_item_1line, users);

        binding.serverUrlEdit.setAdapter(urlAdapter);
        binding.userNameEdit.setAdapter(userAdapter);
    }

    @Override
    public void saveUsersData() {
        if (urls != null && !urls.contains(binding.serverUrlEdit.getText().toString())) {
            urls.add(binding.serverUrlEdit.getText().toString());
            saveListToPreference(Constants.PREFS_URLS, urls);
        }
        if (users != null && !users.contains(binding.userNameEdit.getText().toString())) {
            users.add(binding.userNameEdit.getText().toString());
            saveListToPreference(Constants.PREFS_USERS, users);
        }

        if (false && presenter.canHandleBiometrics() && //TODO: Remove false when green light
                (!BiometricStorage.areCredentialsSet() &&
                        !BiometricStorage.areSameCredentials(
                                binding.serverUrlEdit.getText().toString(),
                                binding.userNameEdit.getText().toString(),
                                binding.userPassEdit.getText().toString()))) {
            showInfoDialog(getString(R.string.biometrics_security_title),
                    getString(R.string.biometrics_security_text),
                    new OnDialogClickListener() {
                        @Override
                        public void onPossitiveClick(AlertDialog alertDialog) {
                            BiometricStorage.saveUserCredentials(
                                    binding.serverUrlEdit.getText().toString(),
                                    binding.userNameEdit.getText().toString(),
                                    binding.userPassEdit.getText().toString());
                            goToNextScreen();
                        }

                        @Override
                        public void onNegativeClick(AlertDialog alertDialog) {
                            goToNextScreen();
                        }
                    }).show();
        } else
            goToNextScreen();

    }


    @Override
    public void onBackPressed() {
        if (isPinScreenVisible) {
            binding.pinLayout.getRoot().setVisibility(View.GONE);
            isPinScreenVisible = false;
        } else {
            super.onBackPressed();
            finish();
        }
    }

    @Override
    public void onSaveInstanceState(Bundle outState, PersistableBundle outPersistentState) {
        super.onSaveInstanceState(outState, outPersistentState);
    }

    @Override
    public void onRestoreInstanceState(Bundle savedInstanceState, PersistableBundle persistentState) {
        super.onRestoreInstanceState(savedInstanceState, persistentState);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == RQ_QR_SCANNER && resultCode == RESULT_OK) {
            qrUrl = data.getStringExtra(Constants.EXTRA_DATA);
        }
    }


}

File: app/src/main/java/org/dhis2/usescases/login/LoginActivity.kt
package org.dhis2.usescases.login

import android.annotation.SuppressLint
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import androidx.activity.SystemBarStyle
import androidx.activity.compose.setContent
import androidx.activity.enableEdgeToEdge
import androidx.compose.ui.graphics.toArgb
import org.dhis2.R
import org.dhis2.bindings.app
import org.dhis2.bindings.buildInfo
import org.dhis2.commons.Constants.SESSION_DIALOG_RQ
import org.dhis2.commons.dialogs.CustomDialog
import org.dhis2.commons.resources.ResourceManager
import org.dhis2.data.server.OpenIdSession
import org.dhis2.mobile.login.authentication.OpenIdController
import org.dhis2.mobile.login.main.ui.navigation.AppLinkNavigation
import org.dhis2.mobile.login.main.ui.screen.LoginScreen
import org.dhis2.usescases.about.PolicyView
import org.dhis2.usescases.general.ActivityGlobalAbstract
import org.dhis2.usescases.main.MainActivity
import org.dhis2.usescases.sync.SyncActivity
import org.hisp.dhis.mobile.ui.designsystem.theme.DHIS2Theme
import org.hisp.dhis.mobile.ui.designsystem.theme.SurfaceColor
import org.koin.android.ext.android.inject
import javax.inject.Inject
import kotlin.getValue

const val EXTRA_SKIP_SYNC = "SKIP_SYNC"
const val EXTRA_ACCOUNT_DISABLED = "EXTRA_ACCOUNT_DISABLED"
const val EXTRA_RENEW_SESSION = "EXTRA_RENEW_SESSION"
const val IS_DELETION = "IS_DELETION"
const val ACCOUNTS_COUNT = "ACCOUNTS_COUNT"
const val FROM_SPLASH = "FROM_SPLASH"

class LoginActivity : ActivityGlobalAbstract() {
    override var handleEdgeToEdge = false

    @Inject
    lateinit var resourceManager: ResourceManager

    private val appLinkNavigation: AppLinkNavigation by inject()
    private val openIdController: OpenIdController by inject()

    private var isPinScreenVisible = false

    private var skipSync = false

    companion object {
        fun bundle(
            skipSync: Boolean = false,
            accountsCount: Int = -1,
            isDeletion: Boolean = false,
            logOutReason: OpenIdSession.LogOutReason? = null,
            fromSplash: Boolean = false,
            renewSession: Boolean = false,
        ): Bundle =
            Bundle().apply {
                putBoolean(EXTRA_SKIP_SYNC, skipSync)
                putBoolean(IS_DELETION, isDeletion)
                putInt(ACCOUNTS_COUNT, accountsCount)
                putBoolean(FROM_SPLASH, fromSplash)
                putBoolean(EXTRA_RENEW_SESSION, renewSession)
                when (logOutReason) {
                    OpenIdSession.LogOutReason.DISABLED_ACCOUNT ->
                        putBoolean(
                            EXTRA_ACCOUNT_DISABLED,
                            true,
                        )

                    null -> {
                        // Nothing to do in this case
                    }
                }
            }
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onCreate(savedInstanceState: Bundle?) {
        enableEdgeToEdge(
            statusBarStyle = SystemBarStyle.dark(SurfaceColor.Primary.toArgb()),
        )
        super.onCreate(savedInstanceState)

        openIdController.bind(this)

        checkMessage()

        skipSync = intent.getBooleanExtra(EXTRA_SKIP_SYNC, false)

        setContent {
            DHIS2Theme {
                LoginScreen(
                    versionName = buildInfo(),
                    onNavigateToHome = {
                        app().createUserComponent()
                        startActivity(MainActivity::class.java, null, true, true, null)
                    },
                    onNavigateToSync = {
                        app().createUserComponent()
                        startActivity(SyncActivity::class.java, null, true, true, null)
                    },
                    onNavigateToPrivacyPolicy = {
                        activity?.let {
                            startActivity(Intent(it, PolicyView::class.java))
                        }
                    },
                    onFinish = {
                        finish()
                    },
                    renewSession = intent.getBooleanExtra(EXTRA_RENEW_SESSION, false),
                )
            }
        }
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        setIntent(intent)
    }

    override fun onResume() {
        super.onResume()
        val appLinkAction = intent?.action
        val appLinkData: Uri? = intent?.data
        if (Intent.ACTION_VIEW == appLinkAction && appLinkData != null) {
            appLinkNavigation.emit(appLinkData.toString())
            intent?.action = null
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        openIdController.unbind()
    }

    @Deprecated("Deprecated in Java")
    override fun onBackPressed() {
        if (isPinScreenVisible) {
            isPinScreenVisible = false
        } else {
            super.onBackPressed()
            finish()
        }
    }

    private fun checkMessage() {
        if (intent.getBooleanExtra(EXTRA_ACCOUNT_DISABLED, false)) {
            showAccountDisabled()
        }
    }

    private fun showAccountDisabled() {
        val sessionDialog =
            CustomDialog(
                this,
                getString(R.string.account_disable_title),
                getString(R.string.account_disable_message),
                getString(R.string.action_accept),
                null,
                SESSION_DIALOG_RQ,
                null,
            )
        sessionDialog.setCancelable(false)
        sessionDialog.show()
    }
}
